### PR TITLE
Feat: add api for app compare or reset 

### PR DESCRIPTION
**Pull request by zxbyoyoyo**
### Description of your changes

Fixes #

I have:

- [ ] Read and followed KubeVela's [contribution process](https://github.com/oam-dev/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/oam-dev/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [ ] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### CODE REVIEW

**Review by barnettZQG — APPROVED** [review]

File: pkg/apiserver/rest/usecase/application.go [hunk 1]
```diff
@@ -1415,3 +1493,220 @@ func genWebhookToken() string {
...
+		DependsOn:        component.DependsOn,
+		Inputs:           component.Inputs,
+		Outputs:          component.Outputs,
+		Scopes:           component.Scopes,
+	}
+	if component.Properties != nil {
+		bc.Properties, _ = model.NewJSONStruct(component.Properties)
+	}
+	for _, trait := range component.Traits {
+		properties, _ := model.NewJSONStruct(trait.Properties)
```
Comment (wonderflow) [inline pkg/apiserver/rest/usecase/application.go:1512]: we should report the error if json marshal failed

File: pkg/apiserver/rest/usecase/application_test.go [hunk 2]
```diff
@@ -506,6 +506,72 @@ var _ = Describe("Test application usecase function", func() {
...
+	It("Test CompareAppWithLatestRevision function", func() {
+
+		appModel, err := appUsecase.GetApplication(context.TODO(), testApp)
+		Expect(err).Should(BeNil())
+		_, err = appUsecase.Deploy(context.TODO(), appModel, v1.ApplicationDeployRequest{WorkflowName: convertWorkflowName("app-dev")})
+		Expect(err).Should(BeNil())
+		component, err := appUsecase.GetApplicationComponent(context.TODO(), appModel, "component-name")
+		Expect(err).Should(BeNil())
+
+		// compare1: when not change, should return false
```
Comment (wonderflow) [inline pkg/apiserver/rest/usecase/application_test.go:518]:
```suggestion
		By("compare when app not change, should return false")
```

File: pkg/apiserver/rest/usecase/application_test.go [hunk 3]
```diff
@@ -506,6 +506,72 @@ var _ = Describe("Test application usecase function", func() {
...
+		compareResponse, err := appUsecase.CompareAppWithLatestRevision(context.TODO(), appModel, v1.AppCompareReq{})
```
Comment (wonderflow) [inline pkg/apiserver/rest/usecase/application_test.go:519]: what do we compare if the compareReq is an empty struct?

File: pkg/apiserver/rest/usecase/application_test.go [hunk 4]
```diff
@@ -506,6 +506,72 @@ var _ = Describe("Test application usecase function", func() {
...
+		Expect(err).Should(BeNil())
+		Expect(cmp.Diff(compareResponse.IsDiff, false)).Should(BeEmpty())
+
+		// compare2: add app's env , not change, should return false
```
Comment (wonderflow) [inline pkg/apiserver/rest/usecase/application_test.go:523]: please use `By` to describe the test

File: pkg/apiserver/rest/usecase/application.go [hunk 5]
```diff
@@ -1344,6 +1354,74 @@ func (c *applicationUsecaseImpl) Statistics(ctx context.Context, app *model.Appl
 	}, nil
 }
 
+// CompareAppWithLatestRevision compare application with last revision
+func (c *applicationUsecaseImpl) CompareAppWithLatestRevision(ctx context.Context, appModel *model.Application, compareReq apisv1.AppCompareReq) (*apisv1.AppCompareResponse, error) {
+	var reqWorkflowName string
+	if compareReq.Env != "" {
+		reqWorkflowName = convertWorkflowName(compareReq.Env)
```
Comment (wonderflow) [inline pkg/apiserver/rest/usecase/application.go:1361]: let's add a test to cover this case, what will happen if the Env is not empty?

**Comment by chivalryq** [comment]
Why should we change the Dockerfile?

File: pkg/apiserver/rest/usecase/application.go [hunk 6]
```diff
@@ -1415,3 +1493,220 @@ func genWebhookToken() string {
...
+			}
+			buff.Write(result)
+			buff.Write([]byte("\n---\n"))
+		}
+		buff.Write([]byte("\n"))
+	}
+	return buff, nil
+}
+
+func ignoreSomeParams(o *v1beta1.Application) {
```
Comment (wonderflow) [inline pkg/apiserver/rest/usecase/application.go:1698]: it seems to be a common function, do we have something similar in `vela dry-run/live-diff` package? Or make it as a utility function

File: pkg/apiserver/rest/usecase/application_test.go [hunk 7]
```diff
@@ -506,6 +506,72 @@ var _ = Describe("Test application usecase function", func() {
...
+	It("Test ResetAppToLatestRevision function", func() {
+		appModel, err := appUsecase.GetApplication(context.TODO(), testApp)
+		Expect(err).Should(BeNil())
+		resetResponse, err := appUsecase.ResetAppToLatestRevision(context.TODO(), testApp)
+		Expect(err).Should(BeNil())
+		Expect(cmp.Diff(resetResponse.IsReset, true)).Should(BeEmpty())
+		component, err := appUsecase.GetApplicationComponent(context.TODO(), appModel, "component-name")
+		Expect(err).Should(BeNil())
+		expectProperties := "{\"image\":\"nginx\"}"
+		Expect(cmp.Diff(component.Properties.JSON(), expectProperties)).Should(BeEmpty())
```
Comment (wonderflow) [inline pkg/apiserver/rest/usecase/application_test.go:572]: please add some more reset cases to cover:

1. change traits
2. change workflows
3. change envs

**Review by wonderflow — COMMENTED** [review]
this is a great feature, let's make it more stable, thanks for your contribution.

File: pkg/apiserver/rest/usecase/application_test.go (hunk 3 above)
Comment (zxbyoyoyo) [reply pkg/apiserver/rest/usecase/application_test.go:519]: 对比场景：用户在UI层修改了App的元数据配置（component,trait,policy，不包含workflow：workflow往往跟发布相关，跟元数据关系不大），页面会有提示【有变更,还未发布】

实现思路：
       1. Env为空：默认跟最后一次发布历史记录中的元数据对比
       2. Env不为空：跟某个环境最后一次发布历史对比;
       因为只对比App元数据还不够直观，所以跟曾庆国讨论了下，先通过DryRun进行渲染，然后在对比渲染后的结果

type AppCompareReq struct {
	Env string `json:"env"`  默认为空
}

Comment (wonderflow) [reply pkg/apiserver/rest/usecase/application_test.go:519]: 嗯，建议加一个 Env 不为空的测试哈

File: pkg/apiserver/rest/usecase/application.go (hunk 6 above)
Comment (zxbyoyoyo) [reply pkg/apiserver/rest/usecase/application.go:1698]: yes,  core code comes from: pkg/plugin/cli/livediff.go:29

**Comment by wonderflow** [comment]
Hi, are you still working on this? This is a great feature, hope we can add more test cases and fix CI. @zxbyoyoyo

**Comment by Somefive** [comment]
Similar problem with @chivalryq. Why do we need to change the Dockerfile? Besides, is there any description or explanation for this PR?

**Comment by zxbyoyoyo** [comment]
> Similar problem with @chivalryq. Why do we need to change the Dockerfile? Besides, is there any description or explanation for this PR?

because compare logic depends on  : references/appfile/dryrun/diff.go:117
by the way , this pr is too long,I will resubmit the pr, this is closed

**Comment by zxbyoyoyo** [comment]
resubmit a new pr, this is closed
